Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift

```diff
@@ -9,201 +9,1069 @@
+                // suspended even though we already cancelled them. We must tolerate this and just resume
+                // the continuation with an error.
+                return .resumeContinuationWithError(
+                    upstreamContinuation: continuation,
+                    error: CancellationError()
+                )
+
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        /// Actions returned by `elementProduced()`.
+        enum ElementProducedAction {
+            /// Indicates that the downstream continuation should be resumed with the element.
+            case resumeContinuation(
+                downstreamContinuation: CheckedContinuation<Element?, Error>,
+                element: Element
+            )
+            /// Indicates that nothing should be done.
+            case none
+        }
+
+        mutating func elementProduced(_ element: Element) -> ElementProducedAction {
+            switch state {
+            case .initial:
+                // Child tasks that are producing elements are only created after we transitioned to `merging`
+                preconditionFailure("Invalid state")
+
+            case let .merging(task, buffer, upstreamContinuations, upstreamsFinished, .some(downstreamContinuation)):
+                // We produced an element and have an outstanding downstream continuation
+                // this means we can go right ahead and resume the continuation with that element
+                precondition(buffer.isEmpty, "We are holding a continuation so the buffer must be empty")
+
+                state = .merging(
+                    task: task,
+                    buffer: buffer,
+                    upstreamContinuations: upstreamContinuations,
+                    upstreamsFinished: upstreamsFinished,
+                    downstreamContinuation: nil
+                )
+
+                return .resumeContinuation(
+                    downstreamContinuation: downstreamContinuation,
+                    element: element
+                )
+
+            case .merging(let task, var buffer, let upstreamContinuations, let upstreamsFinished, .none):
+                // There is not outstanding downstream continuation so we must buffer the element
+                // This happens if we race our upstream sequences to produce elements
+                // and the _losers_ are signalling their produced element
+                state = .modifying
+
+                buffer.append(element)
+
+                state = .merging(
+                    task: task,
+                    buffer: buffer,
+                    upstreamContinuations: upstreamContinuations,
+                    upstreamsFinished: upstreamsFinished,
+                    downstreamContinuation: nil
+                )
+
+                return .none
+
+            case .upstreamThrew:
```

Comment:
small nit: we rarely use past tense for names of cases. `.upstreamFailure` might be a better name.